polaris-react/src/components/DataTable/DataTable.tsx

```diff
@@ -266,15 +421,37 @@ class DataTableInner extends PureComponent<CombinedProps, DataTableState> {
     };
   };
 
-  private scrollListener = () => {
+  private stickyHeaderScrolling = (event: Event) => {
+    const {current: stickyTableHeadingsRow} = this.stickyTableHeadingsRow;
+    const {current: scrollContainer} = this.scrollContainer;
+    const targetIsStickyHeader = event.target === stickyTableHeadingsRow;
+
+    if (stickyTableHeadingsRow == null || scrollContainer == null) {
+      return;
+    }
+
+    const {scrollLeft: stickyScrollLeft} = stickyTableHeadingsRow;
+    const {scrollLeft: scrollContainerScrollLeft} = scrollContainer;
+
+    if (targetIsStickyHeader) {
+      scrollContainer.scrollLeft = stickyScrollLeft;
+    } else {
+      stickyTableHeadingsRow.scrollLeft = scrollContainerScrollLeft;
+    }
+  };
+
+  private scrollListener = (event: Event) => {
     this.setState((prevState) => ({
       ...this.calculateColumnVisibilityData(prevState.condensed),
```

Comment:
This will trigger a re-render of the component including all of the calculations within `calculateColumnVisibilityData` which is partly responsible for the sticky header lagging behind a bit on horizontal scroll.

I would recommend to debounce this specific `setState` call. It's mainly used for the scroll indicators which is fine if only updated every 500ms or so. This way the heavy calculations don't drop the framerate and the sticky header never lags behind.

```tsx
  private scrollListenerStickyHeader = (event: Event) => {
    this.stickyHeaderScrolling(event);
  };

  // eslint-disable-next-line @typescript-eslint/member-ordering <--- somehow eslint doesn't like the debounce, hence that line.
  private scrollListener = debounce(() => {
    this.setState((prevState) => ({
      ...this.calculateColumnVisibilityData(prevState.condensed),
    }));
  });
```

```tsx
<EventListener
    capture
    event="scroll"
    handler={this.scrollListener}
/>
<EventListener
    capture
    event="scroll"
    handler={this.scrollListenerStickyHeader}
/>
```